diff --git i/utils/bazel/llvm-project-overlay/llvm/config.bzl w/utils/bazel/llvm-project-overlay/llvm/config.bzl
index dcb19992321a..a5b388d40289 100644
--- i/utils/bazel/llvm-project-overlay/llvm/config.bzl
+++ w/utils/bazel/llvm-project-overlay/llvm/config.bzl
@@ -28,6 +28,8 @@ def native_arch_defines(arch, triple):
 
 posix_defines = [
     "LLVM_ON_UNIX=1",
+    "HAVE_BACKTRACE=1",
+    "BACKTRACE_HEADER=<execinfo.h>",
     "LLVM_ENABLE_LLVM_EXPORT_ANNOTATIONS=1",
     "LLVM_ENABLE_PLUGINS=1",
     "LLVM_ENABLE_THREADS=1",
@@ -69,24 +71,10 @@ fenv_defines = [
     "HAVE_DECL_FE_INEXACT=1",
 ]
 
-backtrace_defines = select({
-    "@platforms//os:emscripten": [],
-    "@platforms//os:windows": [],
-    "@llvm//platforms/config:musl": [],
-    "//conditions:default": [
-        "HAVE_BACKTRACE=1",
-        "BACKTRACE_HEADER=<execinfo.h>",
-    ],
-})
-
-mallinfo_defines = select({
-    "@llvm//platforms/config:gnu": ["HAVE_MALLINFO=1"],
-    "//conditions:default": [],
-})
-
 linux_defines = posix_so_defines + fenv_defines + [
     "_GNU_SOURCE",
     "HAVE_GETAUXVAL=1",
+    "HAVE_MALLINFO=1",
     "HAVE_SBRK=1",
     "HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC=1",
 ]
@@ -135,7 +123,7 @@ os_defines = select({
     "@platforms//os:macos": macos_defines,
     "@platforms//os:windows": win32_defines,
     "//conditions:default": linux_defines,
-}) + arc4random_defines + backtrace_defines + mallinfo_defines
+}) + arc4random_defines
 
 # HAVE_BUILTIN_THREAD_POINTER is true for on Linux (outside of ppc64) for
 # all recent toolchains. Add it here by default on Linux as we can't perform a
