diff --git a/llvm/include/llvm/CodeGen/MachineFunction.h b/llvm/include/llvm/CodeGen/MachineFunction.h
index a6f4930acb33..f2d50eee13f1 100644
--- a/llvm/include/llvm/CodeGen/MachineFunction.h
+++ b/llvm/include/llvm/CodeGen/MachineFunction.h
@@ -294,7 +294,7 @@ struct LandingPadInfo {
 class LLVM_ABI MachineFunction {
   Function &F;
   const TargetMachine &Target;
-  const TargetSubtargetInfo &STI;
+  const TargetSubtargetInfo *STI;
   MCContext &Ctx;

   // RegInfo - Information about each register in use in the function.
@@ -767,13 +767,13 @@ public:

   /// getSubtarget - Return the subtarget for which this machine code is being
   /// compiled.
-  const TargetSubtargetInfo &getSubtarget() const { return STI; }
+  const TargetSubtargetInfo &getSubtarget() const { return *STI; }

   /// getSubtarget - This method returns a pointer to the specified type of
   /// TargetSubtargetInfo.  In debug builds, it verifies that the object being
   /// returned is of the correct type.
   template<typename STC> const STC &getSubtarget() const {
-    return static_cast<const STC &>(STI);
+    return static_cast<const STC &>(*STI);
   }

   /// getRegInfo - Return information about the registers currently in use.
diff --git a/llvm/lib/CodeGen/MachineFunction.cpp b/llvm/lib/CodeGen/MachineFunction.cpp
index 402937917e6d..27633f7a68c8 100644
--- a/llvm/lib/CodeGen/MachineFunction.cpp
+++ b/llvm/lib/CodeGen/MachineFunction.cpp
@@ -181,7 +181,7 @@ static inline Align getFnStackAlignment(const TargetSubtargetInfo &STI,
 MachineFunction::MachineFunction(Function &F, const TargetMachine &Target,
                                  const TargetSubtargetInfo &STI, MCContext &Ctx,
                                  unsigned FunctionNum)
-    : F(F), Target(Target), STI(STI), Ctx(Ctx) {
+    : F(F), Target(Target), STI(&STI), Ctx(Ctx) {
   FunctionNumber = FunctionNum;
   init();
 }
@@ -212,12 +212,12 @@ void MachineFunction::init() {

   // We can realign the stack if the target supports it and the user hasn't
   // explicitly asked us not to.
-  bool CanRealignSP = STI.getFrameLowering()->isStackRealignable() &&
+  bool CanRealignSP = STI->getFrameLowering()->isStackRealignable() &&
                       !F.hasFnAttribute("no-realign-stack");
   bool ForceRealignSP = F.hasFnAttribute(Attribute::StackAlignment) ||
                         F.hasFnAttribute("stackrealign");
   FrameInfo = new (Allocator) MachineFrameInfo(
-      getFnStackAlignment(STI, F), /*StackRealignable=*/CanRealignSP,
+      getFnStackAlignment(*STI, F), /*StackRealignable=*/CanRealignSP,
       /*ForcedRealign=*/ForceRealignSP && CanRealignSP);
   FrameInfo->setFramePointerPolicy(getFramePointerPolicy(F));

@@ -227,7 +227,7 @@ void MachineFunction::init() {
     FrameInfo->ensureMaxAlignment(*F.getFnStackAlign());

   ConstantPool = new (Allocator) MachineConstantPool(getDataLayout());
-  Alignment = STI.getTargetLowering()->getMinFunctionAlignment();
+  Alignment = STI->getTargetLowering()->getMinFunctionAlignment();

   // -fsanitize=function and -fsanitize=kcfi instrument indirect function calls
   // to load a type hash before the function label. Ensure functions are aligned
@@ -354,7 +354,7 @@ Align MachineFunction::getPreferredAlignment() const {
   if (MaybeAlign A = F.getPreferredAlignment())
     PrefAlignment = *A;
   else if (!F.hasOptSize())
-    PrefAlignment = STI.getTargetLowering()->getPrefFunctionAlignment();
+    PrefAlignment = STI->getTargetLowering()->getPrefFunctionAlignment();
   else
     PrefAlignment = Align(1);

